### ISSUE TITLE
Unification of ability flags and area types

### ISSUE THREAD

**Issue by jakobbotsch** [recastnavigation/recastnavigation#274]
Currently we support 64 area types and 16 ability flags. I have had some gripes with this split when using R&D:
1. Often we want to mark areas with flags too. For example, there might be a city that is cheap to navigate for one kind of agent, but expensive for a different kind. If we have water, ground, road in this city then we would ideally like to mark the whole city with an additional flag, while still retaining the water/ground/road types. But this quickly reduces the amount of area types we have available.
1. 16 ability flags is, in my experience, more than enough. In fact we would often like to trade some of these for more area types.
1. Marking polygons with the correct ability flags is slightly hard because only area types are passed through the entire meshing process. See eg. https://github.com/recastnavigation/recastnavigation/blob/46654531e4e5c17d2ebe1933f3e768f908540f03/RecastDemo/Source/Sample_TileMesh.cpp#L1227 for how we do this right now, but I have had issues with being able to differentiate between different "Ground" polygons to give them correct ability flags.
1. In addition, the split might cause confusion for newer users (I know that I was confused in the beginning!)

I believe we should unify these two concepts to a single, larger variable. I also belive that @memononen did this for Unity's implementation of R&D. Coupled with #260 this will in general give better flexibility in the meshing process.

Now there are a few things to take into account:
1. This will have to be a breaking change, so we will have to bump the mesh version. I don't expect this to be any issue as this will go into 2.0.
1. `dtQueryFilter` will probably have to be made virtual in all cases. I am not quite sure how to write a generic version of it that supports extracting flags/types and calculating costs in a general way. @memononen how was this done in Unity?

I will be starting the work on this very soon.

**Comment by memononen** [comment]
Before you start working on this, I recommend to take a look at how we got there:
http://digestingduck.blogspot.fi/2009/09/area-progress.html
http://digestingduck.blogspot.fi/2009/10/status-update.html
http://digestingduck.blogspot.fi/2010/01/fresh-start-for-areas.html
http://digestingduck.blogspot.fi/2010/02/area-progress-pt-4.html
http://digestingduck.blogspot.fi/2010/02/area-progress-pt-5.html

The area=ability was a bad choice at Unity, done before my time there. That being said, the current situation is not perfect either.

I think it is very important to realise that there are two loosely connected ideas there: 1) travel cost, 2) travel capability. I think these are well reflected in Detour (even if the data type is small), but the build process handles them poorly.

At one point I had automatic prone/crouch/stand area generation working together with area types (grass, road, water). Eventually I ditched it in favor of current implementation because it created so many tiny polygons. And that was just mixing two types. The problem is that once the navmesh has a lot of small polygons, it takes a lot of memory, search comes slow, and the returned paths star to look bad too. There's a tipping point when grids become more feasible.

Just like document access control reflects the organization it belongs to (that is, you cannot create access control system that pleases all organisations), the cost/ability reflects very strongly with the game you're creating. There's no one size fit's all.

I'd suggest to open it up even more, whilst keeping it backwards compatible. Allow the user the define the datatype for area/ability. And allow the user to set a "blend" function for rasterization, which allows to mix the flags in which ever way they choose. Then extend query filter if necessary, and you'll need some changed to dtNavMesh for modifying and getting the values. A default implementation could retain the current functionality.

To sum it up, I think your intent to change the area/flags to behave the same across the board is very good. I think folding cost and ability into one is not a good idea because it generalises too much. Further, the cases which do not fit into the current system well have historically been so tied to game logic that it might be a good idea to see if you could allow the users to fully customize how cost and flags are behave.

**Comment by jakobbotsch** [comment]
I have not looked closely at Unity's implementation, but my intention by folding them into one is to
1. Unite them during the meshing process
1. Allow the user to select exactly how many/which bits he wants to use for abilities and area types

I think this is much better and closer to a one-size-fits-all than the current solution. Especially because we right now end up duplicating a lot of information. For example, when marking a polygon as water, I don't understand why I need to set its flags to swim as well. Instead why do not just have the query filter for a car disallow moving inside "water" polygons, instead of using flags?

Keeping it backwards compatible is hard because of the mixed 6-bit area type/poly type: https://github.com/recastnavigation/recastnavigation/blob/15ebb8bd25e952215d2ff140d5ec01fa50294275/Detour/Include/DetourNavMesh.h#L170

Currently I would like to do the following: https://github.com/jakobbotsch/recastnavigation/blob/265480811e6c30231b78b75d742d7c18ea5150aa/Detour/Include/DetourNavMesh.h#L148
27 bits for `firstLink`, 3 bits for `vertCount` and 2 bits for `type`. Then an additional 32-bits for the unified area/ability flags (it would have been nice to use bit fields, but I believe these are not very portable when serializing to bytes).

I understand why merging lots of small areas might be problematic, and we should not encourage people to do this in the design. However, my current use case is more like the following: we have two factions in our game. Both factions have cities and territories they control. These cities and territories are very large, and might still contain rivers, doors, roads and so on. During meshing we would essentially like to apply a flag that marks a territory as "faction 1" or "faction 2", while still allowing roads/rivers/ground to be marked on top. The ability to do this was already implemented in #260, and after unification it should be much more powerful (since it won't be limited to 6 bits). So the full customization of how cost and flags behave is pretty much already implemented for Recast. We will not be merging tons of areas into other areas and generating a lot of small polys (unless the user really wants to do this) -- it will work just as before for area types, depending on how you specify `rcAreaModification` (as implemented by #260) when marking your volumes. In addition, we also get rid of a step of the build process (where polys are given flags based on their area types in the current implementation).

In Detour, the customization will also be simple. You essentially just implement a custom `dtQueryFilter` that checks the area flags to allow/deny movement (like ability flags were checked before). It also calculates costs depending on the unified field, which can take flags into account as well. For example, coming back to my previous example with two factions in our game: units of both factions can navigate into enemy controlled territories, but would prefer not to. So our implementation of `dtQueryFilter` just needs to look something like the following, for a regular unit:

```
float dtQueryFilter::getCost(const float* pa, const float* pb,
							 const dtPolyRef /*prevRef*/, const dtMeshTile* /*prevTile*/, const dtPoly* /*prevPoly*/,
							 const dtPolyRef /*curRef*/, const dtMeshTile* /*curTile*/, const dtPoly* curPoly,
							 const dtPolyRef /*nextRef*/, const dtMeshTile* /*nextTile*/, const dtPoly* /*nextPoly*/) const
{
	float cost = 2.0f;
	if ((curPoly->area & FACTION_MASK) != m_myFaction)
		cost *= 3;
	switch (curPoly->area & AREA_MASK)
	{
		case AREA_WATER:
			cost *= 2;
			break;
		case AREA_ROAD:
			cost *= 0.5f;
			break;
	}
	return dtVdist(pa, pb) * cost;
}
```

Now lets say we also have cars in the game. Our implementation of `dtQueryFilter::passFilter` for our cars will look something like:
```
bool dtQueryFilter::passFilter(const dtPolyRef /*ref*/,
							   const dtMeshTile* /*tile*/,
							   const dtPoly* poly) const
{
	return (poly->area & AREA_MASK) != AREA_WATER;
}
```

We are in full control of how many bits we want to use on flags (faction controlled) or area types (water, ground, etc.), and there is no duplicated info. Since we only need 1 bit for factions, we get 31 bits for area types. Other people may need way more bits for flags, but they get to choose exactly how they want to make this split, and it is easy to customize it all with `rcAreaModification` during the build process, and query filters when querying. They might also need much more dynamic flags, but they can still just reserve how many bits they want in the unified field.

Overall I believe this makes the whole concept of flags and areas much more "orthogonal" to how you would normally use it, and both simpler to understand and more flexible. The problem with creating too many small polys is definitely something that needs to be considered, but I would rather document this than give up the flexibility of `rcAreaModification`.

**Comment by memononen** [comment]
I had already forgotten the work you've done for `rcAreaModification`, my bad. I think `rcAreaModification` and your use of queryfilter is pretty much exactly what I meant. But I overlooked some of the implementation details (a year and a half of not coding has done it).

I approve your course of action, carry on! :)

**Comment by jakobbotsch** [comment]
So I am a bit unsure what to do with `dtQueryFilter`. The easiest is probably just to make it always virtual. But we could also change the standard implementation to something like this:

```
class dtQueryFilter
{
  std::vector<float> m_areaCost;
  unsigned int m_includeFlags; // Area of polygon must contain one of these bits to be included
  unsigned int m_excludeFlags; // If area of polygon contains one of these bits it is excluded
  unsigned int m_costMask; // Before computing the area cost, the area is anded with this mask

  public:
    dtQueryFilter(unsigned int numAreaTypes)
      : m_areaCost(numAreaTypes)
      , m_includeFlags(0xffffffff)
      , m_excludeFlags(0)
      , m_costMask(dtNextPow2(numAreaTypes) - 1) // Compute mask size from number of area types
    {
      for (unsigned int i = 0; i < numAreaTypes; i++)
        m_areaCost[i] = 1.0f;
    }

    bool passFilter(const dtPolyRef /*ref*/, const dtMeshTile* /*tile*/, const dtPoly* poly) const
    {
        return (poly->area & m_includeFlags) != 0 && (poly->area & m_excludeFlags) == 0;
    }

    float dtQueryFilter::getCost(
        const float* pa, const float* pb,
        const dtPolyRef /*prevRef*/, const dtMeshTile* /*prevTile*/, const dtPoly* /*prevPoly*/,
        const dtPolyRef /*curRef*/, const dtMeshTile* /*curTile*/, const dtPoly* curPoly,
        const dtPolyRef /*nextRef*/, const dtMeshTile* /*nextTile*/, const dtPoly* /*nextPoly*/) const
{
	return dtVdist(pa, pb) * m_areaCost[curPoly->area & m_costMask];
}
};
```

Where the user is allowed to pass in how many area types he wants, and then it implicitly assumes he is using the lower log2(number) bits of the area for this.
Thoughts? Making it always virtual is definitely simpler and makes no assumptions on behalf of the user, but also changes the design a bit compared to how it is right now.

**Comment by hymerman** [comment]
(busy with baby, but...)

I like this. I think making it virtual is fine. I never used the #define to enable virtual filters, I would just edit the source of the base/default filter itself anyway, so I imagine users that care about the virtual call overhead will do the same to remove the virtual-ness, rather than deriving.

**Comment by memononen** [comment]
We had good results at Unity by using virtual functions to do batch processing. IIRC a dozen items was enough to offset the cost for virtual function call. I think the search loop could be arranged to first find usable out going links, then calling cost for all (end of path is going to be a bit tricky), and then processing the rest. Of course, whould be profiled first to see if make any difference.

@hymerman (me too :) a 10 mo old who wants to climb everywhere)

**Comment by jakobbotsch** [comment]
Congrats to both of you!

I suspect that on x86-64 the virtualness is not going to make any substantial difference. The slow part is probably the copying of the args (and we give up a lot of benefit here because of inlining, but even with my changes we would be giving this up). The virtual function tables and functions themselves are going to be very hot in cache during the queries so they should be very fast.
I will do some profiling later and see. There probably is a way to reduce the number of args and the amount of copying done.

**Comment by grahamboree** [comment]
Congrats on the babies!!! 🎉 🍾 👶 🎊 

I like this idea, and it's something that people often ask about as well.  Don't have too much to add beyond what's been said already.  I agree bit-fields are not something we should be using with anything but transient data structures that are accessed simply.  I'm interested to see what profiling info you get about virtual calls in this circumstance as well.  That being said, the default implementation you posted looks fine to me (except for that `std::vector` 😉) and avoids the question of virtual call overhead altogether.

**Comment by jakobbotsch** [comment]
I will close this for now as it turned out to require much more work than I had anticipated. Hopefully I can give this a stab again in the future.
